grakn/client.py

```diff
@@ -149,16 +161,27 @@ def core_client(self, address: ServerAddress) -> _ClientRPC:
     def grakn_cluster_grpc_stub(self, address: ServerAddress) -> GraknClusterStub:
         return self._grakn_cluster_grpc_stubs.get(address)
 
-    def _discover_cluster(self, addresses: List[str]) -> Set[ServerAddress]:
+    def _fetch_cluster_servers(self, addresses: List[str]) -> Set[ServerAddress]:
         for address in addresses:
             try:
                 with _ClientRPC(address) as client:
                     print("Performing cluster discovery to %s..." % address)
                     grakn_cluster_stub = GraknClusterStub(client.channel())
-                    res = grakn_cluster_stub.cluster_discover(cluster_proto.Cluster.Discover.Req())
+                    res = grakn_cluster_stub.cluster_servers(cluster_proto.Cluster.Servers.Req())
                     members = set([ServerAddress.parse(srv) for srv in res.servers])
                     print("Discovered %s" % [str(member) for member in members])
                     return members
             except RpcError:
                 print("Cluster discovery to %s failed." % address)
         raise GraknClientException("Unable to connect to Grakn Cluster. Attempted connecting to the cluster members, but none are available: %s" % str(addresses))
+
+
+class _OpenSessionFailsafeTask(FailsafeTask):
+
+    def __init__(self, database: str, session_type: SessionType, options: GraknClusterOptions, client: "_ClientClusterRPC"):
+        super().__init__(client, database)
+        self.session_type = session_type
+        self.options = options
+
+    def run(self, replica: ReplicaInfo.Replica):
+        return SessionClusterRPC(self.client, replica.address(), self.database, self.session_type, self.options)
```

Comment:
`FailsafeTask` is a new class that holds the failover/retry logic. It is an abstract base class, requiring that `run` is implemented, where `run` represents a unit of work that can fail and should be retried / failed over.